{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from CB_02704 import *\n",
    "path = \"../02704_data\"\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_corr(X, Y): #*\n",
    "    '''Calculates the correlation between rows of masked arrays X and Y'''\n",
    "    \n",
    "    # Annoying-but-needed data setup:\n",
    "    # Make sure arguments are 2-dimensional\n",
    "    X = np.ma.atleast_2d(X)\n",
    "    Y = np.ma.atleast_2d(Y)\n",
    "    \n",
    "    # get the masked arrays as integers\n",
    "    Xnotmask = ~np.ma.getmaskarray(X) * 1\n",
    "    Ynotmask = ~np.ma.getmaskarray(Y) * 1\n",
    "    \n",
    "    # Now onto the calculations:\n",
    "    # Calculate N for each combination of X and Y\n",
    "    N = np.dot(Xnotmask, Ynotmask.T)\n",
    "    \n",
    "    # Calculate the sums\n",
    "    sum_X  = np.ma.dot(X,    Ynotmask.T).filled()\n",
    "    sum_X2 = np.ma.dot(X**2, Ynotmask.T).filled()\n",
    "    \n",
    "    sum_Y  = np.ma.dot(Xnotmask, Y.T).filled()\n",
    "    sum_Y2 = np.ma.dot(Xnotmask, Y.T**2).filled()\n",
    "    \n",
    "    sum_XY = np.ma.dot(X, Y.T * 1.).filled()\n",
    "    \n",
    "    r = ( (N*sum_XY - sum_X*sum_Y)\n",
    "         / np.sqrt(N*sum_X2 - sum_X**2)\n",
    "         / np.sqrt(N*sum_Y2 - sum_Y**2) )\n",
    "         \n",
    "    return r, N"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "ceu_geno = np.load(\"geno_files/ceu_geno.npy\")\n",
    "ceu_geno = np.ma.filled(ceu_geno, 0)\n",
    "\n",
    "yri_geno = np.load(\"geno_files/yri_geno.npy\")\n",
    "yri_geno = np.ma.filled(yri_geno, 0)\n",
    "\n",
    "asw_geno = np.load(\"geno_files/asw_geno.npy\")\n",
    "asw_geno = np.ma.filled(asw_geno, 0)\n",
    "\n",
    "ceu_snp = read_snp_pop(\"ceu\")\n",
    "\n",
    "sim_geno = np.load(\"geno_files/sim_geno.npy\")\n",
    "sim_geno = np.array(sim_geno, dtype=np.int64)\n",
    "snp_ids = ceu_snp.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "r_vals = [0]\n",
    "for i in range(1, len(ceu_geno)):\n",
    "    # print(i)\n",
    "    r, _ = calc_corr(ceu_geno[i-1], ceu_geno[i])\n",
    "    r_vals.append(r[0][0])\n",
    "# print(r_vals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "r_vals_abs = np.array(r_vals)\n",
    "r_vals_abs = np.abs(r_vals_abs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "ceu_snp.index = np.arange(0, len(ceu_snp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "hm3_genetic_map = {\"CHROM\": [], \"POS\": [], \"GENETIC_DIST\": []}\n",
    "\n",
    "dist = 0\n",
    "\n",
    "for i,row in ceu_snp.iterrows():\n",
    "    if row[\"chromosome\"] == 2:\n",
    "        break\n",
    "    hm3_genetic_map[\"CHROM\"].append(row[\"chromosome\"])\n",
    "    hm3_genetic_map[\"POS\"].append(row[\"position\"])\n",
    "    hm3_genetic_map[\"GENETIC_DIST\"].append(0)\n",
    "\n",
    "hm3_genetic_map_df = pd.DataFrame(hm3_genetic_map)\n",
    "hm3_genetic_map_df.to_csv(\"genetic_map.txt\", sep=\"\\t\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GENERATING VCF FILE FOR STACKED CEU AND YRI GENOTYPE DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "header = \"##fileformat=VCFv4.2\\n\"\n",
    "header += \"##fileDate=20241120\\n\"\n",
    "header += \"##FORMAT=<ID=GT,Number=1,Type=String,Description=\\\"Genotype\\\">\\n\"\n",
    "header += \"##INFO=<ID=H3,Number=0,Type=Flag,Description=\\\"HapMap3 membership\\\">\\n\"\n",
    "header += \"#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\"\n",
    "\n",
    "N = np.shape(ceu_geno)[1] + np.shape(yri_geno)[1]\n",
    "for i in range(N):\n",
    "    header += f\"\\tR_SAMPLE{i + 1}\"\n",
    "    \n",
    "stacked_geno = np.ma.hstack((ceu_geno, yri_geno))\n",
    "\n",
    "for i in range(1, 23):\n",
    "    print(f\"Generating VCF file for chromosome {i}\")\n",
    "    vcf_body = \"\"\n",
    "    prev_pos = -1\n",
    "    indices = np.array(ceu_snp[ceu_snp[\"chromosome\"] == i].index)\n",
    "    for j in range(len(indices)):\n",
    "        index = indices[j]\n",
    "        row = ceu_snp.iloc[index]\n",
    "        if int(row['position']) == prev_pos:\n",
    "            continue         \n",
    "        vcf_body += f\"\\n{row['chromosome']}\\t\"\n",
    "        vcf_body += f\"{row['position']}\\t\"\n",
    "\n",
    "        vcf_body += f\"{snp_ids[index]}\\t\"\n",
    "        vcf_body += f\"{row['ref']}\\t\"\n",
    "        vcf_body += f\"{row['alt']}\\t\"\n",
    "        vcf_body += \"100\\t\"\n",
    "        vcf_body += \"PASS\\t\"\n",
    "        vcf_body += \"H3\\t\"\n",
    "        vcf_body += \"GT\\t\"\n",
    "        \n",
    "        stacked_geno[index][np.where(stacked_geno[index] == 9)[0]] = 0\n",
    "\n",
    "        phased_geno = []\n",
    "        for g in stacked_geno[index]:\n",
    "            if g == 0:\n",
    "                phased_geno.append('0|0')\n",
    "            elif g == 1:\n",
    "                phased_geno.append('1|0')\n",
    "            elif g == 2:\n",
    "                phased_geno.append('1|1')\n",
    "        \n",
    "        genotypes = '\\t'.join([g for g in phased_geno])\n",
    "        vcf_body += f\"{genotypes}\"\n",
    "        prev_pos = int(row['position'])\n",
    "\n",
    "\n",
    "    with open(f'vcf_files/chromosome{i}/ceu_yri_geno.vcf', 'w') as f:\n",
    "        f.write(header + vcf_body + \"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GENERATING SAMPLE MAP FOR CEU AND YRI DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"ceu_yri_sample.txt\", \"w\") as f:\n",
    "    N = np.shape(ceu_geno)[1] + np.shape(yri_geno)[1]\n",
    "    for i in range(N):\n",
    "        if i < ceu_geno.shape[1]:\n",
    "            f.write(f\"R_SAMPLE{i+1}\\tCEU\\n\")\n",
    "        elif i < N-1:\n",
    "            f.write(f\"R_SAMPLE{i+1}\\tYRI\\n\")\n",
    "        else:\n",
    "            f.write(f\"R_SAMPLE{i+1}\\tYRI\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### GENERATING VCF FILE FOR ASW GENOTYPE DATA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "header = \"##fileformat=VCFv4.2\\n\"\n",
    "header += \"##fileDate=20241120\\n\"\n",
    "header += \"##FORMAT=<ID=GT,Number=1,Type=String,Description=\\\"Genotype\\\">\\n\"\n",
    "header += \"##INFO=<ID=H3,Number=0,Type=Flag,Description=\\\"HapMap3 membership\\\">\\n\"\n",
    "header += \"#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\"\n",
    "\n",
    "N = np.shape(asw_geno)[1]\n",
    "for i in range(N):\n",
    "    header += f\"\\tQ_SAMPLE{i + 1}\"\n",
    "\n",
    "for i in range(1, 23):\n",
    "    print(f\"Generating VCF file for chromosome {i}\")\n",
    "    vcf_body = \"\"\n",
    "    prev_pos = -1\n",
    "    indices = np.array(ceu_snp[ceu_snp[\"chromosome\"] == i].index)\n",
    "    for j in range(len(indices)):\n",
    "        index = indices[j]\n",
    "        row = ceu_snp.iloc[index]\n",
    "        if int(row['position']) == prev_pos:\n",
    "            continue         \n",
    "        vcf_body += f\"\\n{row['chromosome']}\\t\"\n",
    "        vcf_body += f\"{row['position']}\\t\"\n",
    "\n",
    "        vcf_body += f\"{snp_ids[index]}\\t\"\n",
    "        vcf_body += f\"{row['ref']}\\t\"\n",
    "        vcf_body += f\"{row['alt']}\\t\"\n",
    "        vcf_body += \"100\\t\"\n",
    "        vcf_body += \"PASS\\t\"\n",
    "        vcf_body += \"H3\\t\"\n",
    "        vcf_body += \"GT\\t\"\n",
    "        \n",
    "        asw_geno[index][np.where(asw_geno[index] == 9)[0]] = 0\n",
    "\n",
    "        phased_geno = []\n",
    "        for g in asw_geno[index]:\n",
    "            if g == 0:\n",
    "                phased_geno.append('0|0')\n",
    "            elif g == 1:\n",
    "                phased_geno.append('1|0')\n",
    "            elif g == 2:\n",
    "                phased_geno.append('1|1')\n",
    "        \n",
    "        genotypes = '\\t'.join([g for g in phased_geno])\n",
    "        vcf_body += f\"{genotypes}\"\n",
    "        prev_pos = int(row['position'])\n",
    "\n",
    "\n",
    "    with open(f'vcf_files/chromosome{i}/asw_geno.vcf', 'w') as f:\n",
    "        f.write(header + vcf_body + \"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "header = \"##fileformat=VCFv4.2\\n\"\n",
    "header += \"##fileDate=20241120\\n\"\n",
    "header += \"##FORMAT=<ID=GT,Number=1,Type=String,Description=\\\"Genotype\\\">\\n\"\n",
    "header += \"##INFO=<ID=H3,Number=0,Type=Flag,Description=\\\"HapMap3 membership\\\">\\n\"\n",
    "header += \"#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\tFORMAT\"\n",
    "\n",
    "N = np.shape(sim_geno)[1]\n",
    "for i in range(N):\n",
    "    header += f\"\\tQ_SAMPLE{i + 1}\"\n",
    "\n",
    "for i in range(1, 23):\n",
    "    print(f\"Generating VCF file for chromosome {i}\")\n",
    "    vcf_body = \"\"\n",
    "    prev_pos = -1\n",
    "    indices = np.array(ceu_snp[ceu_snp[\"chromosome\"] == i].index)\n",
    "    \n",
    "    for j in range(len(indices)):\n",
    "        index = indices[j]\n",
    "        row = ceu_snp.iloc[index]\n",
    "        if int(row['position']) == prev_pos:\n",
    "            continue         \n",
    "        vcf_body += f\"\\n{row['chromosome']}\\t\"\n",
    "        vcf_body += f\"{row['position']}\\t\"\n",
    "\n",
    "        vcf_body += f\"{snp_ids[index]}\\t\"\n",
    "        vcf_body += f\"{row['ref']}\\t\"\n",
    "        vcf_body += f\"{row['alt']}\\t\"\n",
    "        vcf_body += \"100\\t\"\n",
    "        vcf_body += \"PASS\\t\"\n",
    "        vcf_body += \"H3\\t\"\n",
    "        vcf_body += \"GT\\t\"\n",
    "        \n",
    "        sim_geno[index][np.where(sim_geno[index] == 9)[0]] = 0\n",
    "\n",
    "        phased_geno = []\n",
    "        for g in sim_geno[index]:\n",
    "            if g == 0:\n",
    "                phased_geno.append('0|0')\n",
    "            elif g == 1:\n",
    "                phased_geno.append('1|0')\n",
    "            elif g == 2:\n",
    "                phased_geno.append('1|1')\n",
    "        \n",
    "        genotypes = '\\t'.join([g for g in phased_geno])\n",
    "        vcf_body += f\"{genotypes}\"\n",
    "        prev_pos = int(row['position'])\n",
    "\n",
    "\n",
    "    with open(f'vcf_files/chromosome{i}/sim_geno.vcf', 'w') as f:\n",
    "        f.write(header + vcf_body + \"\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
